File: test_token.py
"""
test_token.py - personal dev script, not intended for end users.

Checks that tokens are valid and data is fetchable for both Google Health
and Withings providers. Prints a brief summary per provider.

Requires the Python package to be installed:
    cd python && pip install -e .

Set TOKENBRIDGE_URL and TOKENBRIDGE_API_KEY in .env before running.
"""

import sys
import os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), "python"))

from datetime import date, datetime, timedelta, timezone
from tokenbridge import TokenBridge

USER_ID = "jackmanners"
END     = date.today().isoformat()
START   = (date.today() - timedelta(days=7)).isoformat()

tb = TokenBridge()
print(f"TokenBridge: {tb.url}")
print(f"User:        {USER_ID}")
print(f"Window:      {START} → {END}\n")

for provider, data_type in (("google-health", "sleep"), ("withings", "sleep-summary")):
    try:
        records = tb.fetch(USER_ID, data_type, START, END, provider=provider)
        print(f"[OK]   {provider:20s}  {len(records)} records")

        if records:
            if provider == "withings":
                latest = max(records, key=lambda r: r.get("startdate", 0))
                start  = datetime.fromtimestamp(latest["startdate"], timezone.utc).strftime("%Y-%m-%d %H:%M")
                end    = datetime.fromtimestamp(latest["enddate"],   timezone.utc).strftime("%Y-%m-%d %H:%M")
                d      = latest.get("data") or {}
                mins    = d.get("total_sleep_time")
                ahi    = d.get("apnea_hypopnea_index")
                print(f"       latest:   {start} → {end}")
                print(f"       asleep: {f'{mins//3600}h {(mins%3600)//3600}m' if mins is not None else 'n/a'}"
                      f"   AHI: {ahi if ahi is not None else 'n/a'}")
            else:
                latest = max(records, key=lambda r: r.get("sleep.interval.endTime", ""))
                start  = latest.get("sleep.interval.startTime", "n/a")
                end    = latest.get("sleep.interval.endTime",   "n/a")
                mins   = latest.get("sleep.summary.minutesAsleep")
                print(f"       latest:   {start} → {end}")
                print(f"       asleep:   {f'{int(mins)//60}h {int(mins)%60}m' if mins else 'n/a'}")

    except Exception as e:
        print(f"[FAIL] {provider:20s}  {e}")
        if hasattr(e, "response") and e.response is not None:
            print(f"       {e.response.text[:300]}")
    print()

File: test_token.R
# test_token.R - personal dev script, not intended for end users.
#
# Checks that tokens are valid and data is fetchable for both Google Health
# and Withings providers. Prints a brief summary per provider.
#
# Requires the R package to be installed:
#   devtools::install_local("r", force = TRUE)
#
# Set TOKENBRIDGE_URL and TOKENBRIDGE_API_KEY in .env before running.

library(tokenbridge)

USER_ID <- "jackmanners"
END     <- format(Sys.Date(), "%Y-%m-%d")
START   <- format(Sys.Date() - 7, "%Y-%m-%d")

cat("TokenBridge:", Sys.getenv("TOKENBRIDGE_URL"), "\n")
cat("User:       ", USER_ID, "\n")
cat("Window:     ", START, "→", END, "\n\n")

checks <- list(
  list(provider = "google-health", data_type = "sleep"),
  list(provider = "withings",      data_type = "sleep-summary")
)

for (check in checks) {
  provider  <- check$provider
  data_type <- check$data_type

  tryCatch({
    tb_set_provider(provider)
    records <- tb_fetch(USER_ID, data_type, START, END)
    n       <- nrow(records)
    cat(sprintf("[OK]   %-20s  %d records\n", provider, n))

    if (n > 0) {
      if (provider == "withings") {
        latest <- records[which.max(records$startdate), ]
        start  <- format(as.POSIXct(latest$startdate, origin = "1970-01-01", tz = "UTC"), "%Y-%m-%d %H:%M")
        end    <- format(as.POSIXct(latest$enddate,   origin = "1970-01-01", tz = "UTC"), "%Y-%m-%d %H:%M")
        mins   <- latest$data[[1]]$total_sleep_time
        ahi    <- latest$data[[1]]$apnea_hypopnea_index
        cat(sprintf("       latest:   %s → %s\n", start, end))
        cat(sprintf("       asleep:   %s   AHI: %s\n",
                    if (!is.null(mins) && !is.na(mins))
                      sprintf("%dh %dm", mins %/% 3600L, (mins %% 3600L) %/% 60L)
                    else "n/a",
                    if (!is.null(ahi) && !is.na(ahi)) ahi else "n/a"))
      } else {
        latest <- records[which.max(records[["sleep.interval.endTime"]]), ]
        start  <- latest[["sleep.interval.startTime"]]
        end    <- latest[["sleep.interval.endTime"]]
        mins   <- suppressWarnings(as.integer(latest[["sleep.summary.minutesAsleep"]]))
        cat(sprintf("       latest:   %s → %s\n", start, end))
        cat(sprintf("       asleep:   %s\n",
                    if (!is.na(mins)) sprintf("%dh %dm", mins %/% 60L, mins %% 60L)
                    else "n/a"))
      }
    }

  }, error = function(e) {
    cat(sprintf("[FAIL] %-20s  %s\n", provider, conditionMessage(e)))
  })

  cat("\n")
}
